File: morphac/utils/test/transforms_utils_test.cc
#include "gtest/gtest.h"

#include "utils/include/transforms_utils.h"

namespace {

using std::sqrt;

using Eigen::MatrixXd;
using Eigen::Vector2d;
using Eigen::VectorXd;

using morphac::utils::HomogenizePoints;
using morphac::utils::RotatePoints;
using morphac::utils::TransformPoints;
using morphac::utils::UnHomogenizePoints;

TEST(TransformsUtilsTest, HomogenizePoints) {
  srand(7);
  Points points = Points::Random(10, 2);

  HomogeneousPoints homogenous_points = HomogenizePoints(points);

  ASSERT_TRUE(homogenous_points.col(0).isApprox(points.col(0)));
  ASSERT_TRUE(homogenous_points.col(1).isApprox(points.col(1)));
  ASSERT_TRUE(homogenous_points.col(2).isApprox(VectorXd::Ones(10)));
}

TEST(TransformsUtilsTest, UnHomogenizePoints) {
  srand(7);
  Points points = Points::Random(20, 2);

  HomogeneousPoints homogenous_points = HomogenizePoints(points);
  Points recovered_points = UnHomogenizePoints(homogenous_points);

  ASSERT_TRUE(recovered_points.isApprox(points));
}

TEST(TransformsUtilsTest, TransformPoints) {
  // A square polygon for testing.
  MatrixXd points(4, 2), rotated_points(4, 2), translated_points(4, 2),
      transformed_points(4, 2);
  points << -2, 2, 2, 2, 2, -2, -2, -2;
  rotated_points << 0, 2 * sqrt(2), 2 * sqrt(2), 0, 0, -2 * sqrt(2),
      -2 * sqrt(2), 0;
  translated_points << 1, 4, 5, 4, 5, 0, 1, 0;
  transformed_points << 0, 7 + 2 * sqrt(2), 2 * sqrt(2), 7., 0, 7 - 2 * sqrt(2),
      -2 * sqrt(2), 7.;

  // Test identity transformation.
  ASSERT_TRUE(TransformPoints(points, 0., Vector2d::Zero()).isApprox(points));

  // Test rotation.
  ASSERT_TRUE(TransformPoints(points, -M_PI / 4, Vector2d::Zero())
                  .isApprox(rotated_points));
  // Invariance.
  ASSERT_TRUE(
      TransformPoints(TransformPoints(points, M_PI / 3, Vector2d::Zero()),
                      -M_PI / 3, Vector2d::Zero())
          .isApprox(points));

  // Test translation.
  ASSERT_TRUE(TransformPoints(points, 0., Vector2d{3., 2.})
                  .isApprox(translated_points));
  // Invariance.
  ASSERT_TRUE(TransformPoints(TransformPoints(points, 0., Vector2d{-8., 13.}),
                              0., Vector2d{8., -13.})
                  .isApprox(points));

  // Transformation.
  ASSERT_TRUE(TransformPoints(points, -M_PI / 4, Vector2d{0., 7.})
                  .isApprox(transformed_points));
  // Invariance.
  //ASSERT_TRUE(
  //    TransformPoints(TransformPoints(points, -M_PI / 4, Vector2d{0., 7.}),
  //                    M_PI / 4, Vector2d{0., -7.})
  //        .isApprox(points));
}

}  // namespace

int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: morphac/utils/python/test/transforms_utils_test.py
import numpy as np
import pytest

from morphac.utils import (
    homogenize_points,
    rotate_points,
    transform_points,
    translate_points,
    unhomogenize_points,
)


@pytest.fixture
def generate_points_list():
    np.random.seed(7)
    points = np.random.randn(10, 2)
    homogeneous_points = np.random.randn(10, 3)

    return points, homogeneous_points


# As the cpp test does the numeric computation, here we only do some basic
# testing to ensure that the binding interface works correctly.


def test_homogenize_points(generate_points_list):

    points, _ = generate_points_list

    homogeneous_points = homogenize_points(points=points)

    assert isinstance(homogeneous_points, np.ndarray)
    assert homogeneous_points.dtype == np.float64

    assert homogeneous_points.shape == (10, 3)

    assert np.allclose(points, homogeneous_points[:, :2])
    assert np.allclose(
        homogeneous_points[:, -1], np.ones(homogeneous_points.shape[0]))


def test_invalid_homogenize_points(generate_points_list):

    _, homogeneous_points = generate_points_list

    # Calling the function with an array of incompatible size.
    with pytest.raises(TypeError):
        _ = homogenize_points(points=homogeneous_points)


def test_unhomogenize_points(generate_points_list):

    _, homogeneous_points = generate_points_list

    points = unhomogenize_points(homogeneous_points=homogeneous_points)

    assert isinstance(points, np.ndarray)
    assert points.dtype == np.float64

    assert points.shape == (10, 2)

    assert np.allclose(points, homogeneous_points[:, :2])


def test_invalid_unhomogenize_points(generate_points_list):

    points, _ = generate_points_list

    # Calling the function with an array of incompatible size.
    with pytest.raises(TypeError):
        _ = unhomogenize_points(homogeneous_points=points)


def test_translate_points(generate_points_list):

    points, _ = generate_points_list

    translated_points = translate_points(points=points, translation=[1., -2.])

    assert isinstance(translated_points, np.ndarray)
    assert translated_points.dtype == np.float64

    assert translated_points.shape == points.shape


def test_invalid_translate_points(generate_points_list):

    points, homogeneous_points = generate_points_list

    with pytest.raises(TypeError):
        _ = translate_points(homogeneous_points, [0., 0.])
    with pytest.raises(TypeError):
        _ = translate_points(points, [0.])
    with pytest.raises(TypeError):
        _ = translate_points(points, [0., 0., 0.])


def test_rotate_points(generate_points_list):

    points, _ = generate_points_list

    rotated_points = rotate_points(points=points, angle=np.pi / 3.)

    assert isinstance(rotated_points, np.ndarray)
    assert rotated_points.dtype == np.float64

    assert rotated_points.shape == points.shape


def test_invalid_rotate_points(generate_points_list):

    points, homogeneous_points = generate_points_list

    with pytest.raises(TypeError):
        _ = rotate_points(homogeneous_points, 0.)


def test_transform_points(generate_points_list):

    points, _ = generate_points_list

    transformed_points = transform_points(
        points=points, angle=np.pi / 3., translation=[1., -2.])

    assert isinstance(transformed_points, np.ndarray)
    assert transformed_points.dtype == np.float64

    assert transformed_points.shape == points.shape


def test_invalid_transform_points(generate_points_list):

    points, homogeneous_points = generate_points_list

    with pytest.raises(TypeError):
        _ = transform_points(homogeneous_points, 0., [0., 0.])
